[Merged by Bors] - feat(algebra/algebra/spectrum): define spectrum and prove basic prope...

j-loreaux opened the pull request:
…rties

Define the resolvent set and the spectrum of an element of an algebra as
a set of elements in the scalar ring. We prove a few basic facts
including that additive cosets of the spectrum commute with the
spectrum, that is, r + σ a = σ (r ⬝ 1 + a). Similarly, multiplicative
cosets of the spectrum also commute when the element r is a unit of
the scalar ring R. Finally, we also show that the units of R in
σ (a*b) coincide with those of σ (b*a).



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux committed:
feat(algebra/algebra/spectrum): define spectrum and prove basic properties

Define the resolve set and the spectrum of an element of an algebra as
a set of elements in the scalar ring. We prove a few basic facts
including that additive cosets of the spectrum commute with the
spectrum, that is, r + σ a = σ (r ⬝ 1 + a). Similarly, multiplicative
cosets of the spectrum also commute when the element r is a unit of
the scalar ring R. Finally, we also show that the units of R in
σ (a*b) coincide with those of σ (b*a).

j-loreaux committed:
import only tactic.noncomm_ring

fix the lint style error about importing the
entire tactic folder.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+local notation `σ` := spectrum
+
+lemma mem_spectrum_iff_not_unit {r : R} {a : A} :
+  r ∈ (σ a : set R) ↔ ¬ is_unit (r • 1 - a) :=
```
The fact that we have to systematically specify `set R` here means that we probably want `R` to be an explicit argument to `resolvent` and `spectrum`. Then, you can change the local notation to have `σ` mean `spectrum R`, and the problem goes away for the rest of the file.

j-loreaux replied:
Interesting, I had tried that at one point, but for some reason it didn't work for me. I probably had some other issue and only thought that this was the problem.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+  r • s ∈ (σ (r • a) : set R) ↔ s ∈ (σ a : set R) :=
+begin
+  apply not_iff_not.mpr,
+  change is_unit ((r•s)•1 - r•a) ↔ is_unit (s•1 - a),
```
Having to do this `change` trick multiple times indicates that we actually want a lemma for this, in this case something like `mem_resolvent_iff` or something like that.

j-loreaux replied:
yes, I thought something like that would be the right way to go, but wasn't sure exactly how to do it. So the idiom for iff statements in the spectrum would then be `rw [not_iff_not, mem_resolvent_iff]`, or something.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+end
+
+theorem left_add_coset_spectrum (a : A) (r : R) :
+  left_add_coset r (σ a) = (σ (r • 1 + a)) :=
```
```suggestion
  left_add_coset r (σ a) = σ (r • 1 + a) :=
```

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+    have inv_smul_eq : r⁻¹•(r•s•1 - r•a) = s•1 - a,
+      by simp [smul_sub, smul_smul],
+    rw ←inv_smul_eq,
+    exact (unit_mul_unit r⁻¹ h'.unit).is_unit, },
```
```suggestion
    exact (r⁻¹ • h'.unit).is_unit, },
```

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+    exact (unit_mul_unit r⁻¹ h'.unit).is_unit, },
+  { intro h',
+    rw ←smul_sub,
+    exact (unit_mul_unit r h'.unit).is_unit, },
```
```suggestion
    exact (r • h'.unit).is_unit, },
```

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+     exact ⟨⟨1-b*a,1+b*h.unit.inv*a,h₁,h₂⟩,rfl⟩, }
+
+lemma is_unit_one_sub_mul_iff_swap {a b : A} :
+  is_unit (1 - a * b) ↔ is_unit (1 - b * a) :=
```
For something like this that's basically just one proof step, it's probably better to just prove it locally inside the relevant proof with `have`. The same goes for `right_inv_of_right_inv_swap` and so on.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+  have with_smul_one : is_unit (r•(1 : R)•1 - a) ↔ is_unit ((1 : R)•1 - r⁻¹•a),
+    by exact is_unit_smul_smul_sub_iff_is_unit_smul_sub_smul,
+  simp at with_smul_one,
+  exact with_smul_one,
```
```suggestion
  simpa using with_smul_one
```

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+ by {simp [smul_mul_smul]}⟩
+
+lemma is_unit_smul_smul_sub_smul_iff {r : units R} {s : R} {a : A} :
+  is_unit (r • s • 1 - r • a) ↔ is_unit (s • 1 - a) :=
```
```suggestion
lemma is_unit.smul_smul_sub_smul_iff {r : units R} {s : R} {a : A} :
  is_unit (r • s • 1 - r • a) ↔ is_unit (s • 1 - a) :=
```
This seems a bit too specialized to be in the root namespace, so I would "hide it" a bit in the `is_unit` namespace.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+begin
+  apply not_iff_not.mpr,
+  change is_unit (r•1 - a*b) ↔ is_unit (r•1 - b*a),
+  repeat {rw [is_unit_smul_sub_iff_is_unit_sub_smul]},
```
You can probably replace repeat rw by `simp only` -- that's basically what it does.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+⟨r•↑a,
+ r⁻¹•↑a⁻¹,
+ by {simp [smul_mul_smul]},
+ by {simp [smul_mul_smul]}⟩
```
```suggestion
```
There is already a smul operation defined on units, so this definition is not needed.

dupuisf reviewed (commented):
Thanks, this looks pretty good! I left some comments, but I think the main issue left is the question of namespacing; as it stands right now, everything is in the root namespace. Usually we want to reserve this for the really important definitions and lemmas, so the candidates here would be `spectrum` and `resolvent` (and even then, I'm not sure spectrum won't collide with something else...), and then the lemmas that pertain to those should be in the `spectrum` or `resolvent` namespace.

j-loreaux committed:
remove unit_mul_unit and references to it

there is already an instance of `has_scalar (units R) (units A)`, so it was not necessary

Co-authored-by: Frédéric Dupuis <31101893+dupuisf@users.noreply.github.com>

j-loreaux committed:
avoid explicit typing of spectrum

pass the scalar ring as an explicit type to the resolvent
and spectrum functions, and then include it in the notation
for the spectrum. This way Lean can properly find the type
of the spectrum (since A may be an algebra over various
rings). Also delete references to the explicit type of the
spectrum throughout.

j-loreaux committed:
hide some is_unit iffs in the is_unit namespace

j-loreaux committed:
pull swap lemmas into the proof of swap theorem

j-loreaux committed:
improve a few lean idioms

j-loreaux committed:
replace unnecessary changes with an iff lemma

j-loreaux committed:
put everything in the spectrum namespace

we might need to change this namespace at some point. Almost certainly
it will eventually collide with another kind of spectrum.

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+is the `set R` consisting of those `r : R` for which `r•1 - a` is a unit of the
+algebra `A`.  -/
+definition resolvent (a : A) : set R :=
+{ r : R | is_unit (r • 1 - a) }
```
I think we usually spell this
```suggestion
{ r : R | is_unit (algebra_map R A r - a) }
```

j-loreaux replied:
Is this something I should change throughout, or just here? It seems a bit cumbersome to have everywhere.

eric-wieser replied:
Yes, it would be throughout

eric-wieser replied:
Perhaps wait for the opinion of a second reviewer

dupuisf replied:
To me `r • 1` looks a lot nicer, so if there's no big downside (like the API being more developed for that version or something like that) I would vote for that.

j-loreaux replied:
I made the suggested change. With the notation Eric suggested in Zulip, it really wasn't too bad.

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+
+lemma mem_spectrum_iff_not_unit {r : R} {a : A} :
+  r ∈ σ a ↔ ¬ is_unit (r • 1 - a) :=
+by { tidy }
```
Does this work?
```suggestion
iff.rfl
```

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+and `a : units A`, then `unit_mul_unit r a` constructs a `units A` with value `r•a`. -/
+
+lemma is_unit.smul_smul_sub_smul_iff {r : units R} {s : R} {a : A} :
+  is_unit (r • s • 1 - r • a) ↔ is_unit (s • 1 - a) :=
```
I think there's a more general lemma hiding here along the lines of
```lean
  is_unit (r • a) ↔ is_unit a :=
```

j-loreaux replied:
Yes, good point. I proved that and removed several of the more specific lemmas that were only used in a few places.

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+
+-- `r ∈ σ(a*b) ↔ r ∈ σ(b*a)` for any `r : units R`
+theorem unit_mem_spectrum_mul_iff_swap_mul {a b : A} {r : units R} :
+  ↑r ∈ σ (a * b) ↔ ↑r ∈ σ (b * a) :=
```
I think this might be better as
```suggestion
  σ (a * b) = σ (b * a) :=
```

eric-wieser replied:
Oh nevermind, that's false! I meant
```suggestion
  (coe : units R → R) ⁻¹' σ (a * b) = coe ⁻¹'  σ (b * a) :=
```
In which case it probably makes sense to keep both versions

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+  simp [h_eq],
+end
+
+lemma mul_mem_spectrum_mul {a : A} {s : R} {r : units R} :
```
```suggestion
lemma smul_mem_spectrum_smul {a : A} {s : R} {r : units R} :
```

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+end
+
+lemma mul_mem_spectrum_mul {a : A} {s : R} {r : units R} :
+  r • s ∈ σ (r • a) ↔ s ∈ σ a :=
```
Is the following true? (with `open_locale pointwise`)
```lean
   σ (r • a) = r • σ a :=
```

j-loreaux replied:
Yes, this is nice. I didn't realize we had access to such nice ways of stating this.

alexjbest commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *resolvent* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is a unit of the
+algebra `A`.  -/
+definition resolvent (a : A) : set R :=
```
```suggestion
def resolvent (a : A) : set R :=
```

alexjbest commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,187 @@
+  apply not_iff_not.mpr,
+  simp only [mem_resolvent_iff],
+  have h_eq : (r+s)•1 - (s•1 + a) = r•1 - a,
+    by { rw add_smul, noncomm_ring },
```
```suggestion
  { rw add_smul, noncomm_ring },
```

j-loreaux committed:
simplify proofs

j-loreaux committed:
provide a set version of σ(a*b) = σ(b*a) for units

j-loreaux committed:
improve naming

j-loreaux committed:
prove σ (r • a) = r • σ a when r : units R

j-loreaux committed:
remove extraneous documentation

j-loreaux committed:
prove more general is_unit (r ⬝ a) ↔ is_unit a

allows the removal of some lemmas which were too
specific.

j-loreaux committed:
improve naming

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,195 @@
+
+-- products of scalar units and algebra units
+
+lemma is_unit.smul_iff {r : units R} {a : A} :
```
Is this true for any (comm?) group `G`?

```suggestion
lemma is_unit.smul_iff [group G] [mul_action G A] [smul_comm_class G A A] [is_scalar_tower G A A] {r : G} {a : A} :
```

dupuisf replied:
This is one example of the namespace issue: right now the full name of this lemma is `spectrum.is_unit.smul_iff`; I think it should just be `is_unit.smul_iff`.

j-loreaux replied:
yes, this works, I'll address the namespace issue too.

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,195 @@
+    { intro h, let a' : units A :=
+        ⟨a,
+         r•↑(h.unit)⁻¹,
+         by rw [mul_smul_comm, ←smul_mul_assoc, is_unit.mul_coe_inv],
```
Does the `smul_mul_smul` lemma help here?

j-loreaux replied:
not in a way that I can work out, but maybe I'm just naive/blind to it.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,195 @@
+
+universes u v
+
+namespace spectrum
```
I've thought a bit more about the namespace issue, and it looks like the main "competitor" is already called `prime_spectrum`, so it's probably not too crazy to put `spectrum` directly in the root namespace. It looks like `resolvent` should be OK too -- perhaps we could call it `resolvent_set` to make sure. Maybe we could ask on Zulip and see if anyone complains.

Given this though, what goes in the `spectrum` namespace is still on a case-by-case basis, I don't think it makes sense to put everything in this file under that namespace, especially if it leads to duplication of names (i.e. we could have either `mem_spectrum_iff` or `spectrum.mem_iff`, but `spectrum.mem_spectrum_iff` seems unnecessarily long). Also, the lemmas pertaining to `is_unit` shouldn't be under `spectrum` either (see my other comments).

j-loreaux replied:
I think I have addressed all the name/namespace issues now. Declarations in the `spectrum` namespace no longer include that word in their names, `spectrum` and `resolvent` both appear in the root namespace, and the two `is_unit` lemmas are no longer in the `spectrum` namespace.

dupuisf commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,195 @@
+      exact ⟨a',rfl⟩, },
+end
+
+lemma is_unit.smul_sub_iff_sub_inv_smul {r : units R} {a : A} :
```
This should also not be under `spectrum`.

dupuisf reviewed (commented):
Thanks, I think we're almost there! I still left some comments regarding the namespace issue but otherwise I'm very happy with this.

j-loreaux committed:
generalize is_unit.smul_iff

j-loreaux committed:
remove unnecessary braces

kim-em commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,195 @@
+* `resolvent a`: has type `set R` is the resolvent set of an element `a : A` where
+  `A` is an  `R`-algebra.
+* `spectrum a`: has type `set R` is the spectrum of an element `a : A` where
+  `A` is an  `R`-algebra.
```
```suggestion
* `resolvent a : set R`: the resolvent set of an element `a : A` where
  `A` is an  `R`-algebra.
* `spectrum a : set R`: the spectrum of an element `a : A` where
  `A` is an  `R`-algebra.
```

kim-em commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,195 @@
+
+## Main statements
+
+* `unit_left_coset_spectrum`: units in the scalar ring commute (multplication) with the spectrum.
```
```suggestion
* `unit_left_coset_spectrum`: units in the scalar ring commute (multiplication) with the spectrum.
```

j-loreaux committed:
remove left coset spectrum theorem

This is a duplicate after we added the
`smul_eq_smul` theorem.

j-loreaux committed:
address naming and namespace issues

Keep the two is_unit lemmas out of the spectrum
namespace, put the `spectrum` and `resolvent`
definitions in the root namespace. Change names
of declarations to not repeat spectrum (e.g., use
`mem_iff` instead of `mem_iff_spectrum` since it
is already in the `spectrum` namespace).

j-loreaux committed:
fix lint error: unused argument

j-loreaux committed:
improve header comment

fix typos, descriptions of the `resolvent` and
`spectrum` definitions, and references to the
names of the main statements

j-loreaux committed:
change r • 1 to algebra_map R A r throughout

use the notation `↑ₐ` for `algebra_map R A`

dupuisf reviewed (approved):
Thanks -- this looks good to me!

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+
+-- products of scalar units and algebra units
+
+lemma is_unit.smul_iff {G : Type u} [group G] [mul_action G A]
```
This lemma does not belong in this file, and should be called `is_unit_smul_iff`. Could you make another PR to move it to a better place, probably in the `group_action` folder?

I wasn't expecting someone else to merge this so quickly!

j-loreaux replied:
sure, no problem

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+lemma is_unit.smul_sub_iff_sub_inv_smul {r : units R} {a : A} :
+  is_unit (r • 1 - a) ↔ is_unit (1 - r⁻¹ • a) :=
+begin
+  have a_eq : a = r•r⁻¹•a, by simp,
```
This is a great technique for seeing if mathlib can solve the problem, but is the type of thing you should normally clean up once you PR to mathlib.

In particular, if you do `by simp?` you'll find that we already have a lemma for this, and it's called `smul_inv_smul`; so you can replace `a_eq` below with `smul_inv_smul`.

eric-wieser commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -0,0 +1,183 @@
+begin
+  apply not_iff_not.mpr,
+  simp only [mem_resolvent_iff, algebra.algebra_map_eq_smul_one],
+  have h_eq : (r•s)•(1 : A) = r•s•1, by simp,
```
As above; `by simp?` will tell you this is just `smul_assoc`.
